cpp/src/strings/convert/convert_urls.cu

```diff
@@ -185,137 +193,198 @@ constexpr bool is_hex_digit(char ch)
-   * output. If the position is the start of an escape sequence then the sequence is decoded to
-   * produce the character copied to the output. Any characters after the start of an escape
-   * sequence but still within the escape sequence are discarded.
-   *
-   * @param input_idx The input character position to process
-   */
-  __device__ void operator()(size_type input_idx) const
-  {
-    char ch = d_in_chars[input_idx];
-
-    // determine the number of escape sequences at or before this character position
-    size_type const* next_esc_pos_ptr =
-      thrust::upper_bound(thrust::seq, d_esc_positions.begin(), d_esc_positions.end(), input_idx);
-    size_type num_prev_esc = next_esc_pos_ptr - d_esc_positions.data();
-
-    // every escape that occurs before this one replaces 3 characters with 1
-    size_type output_idx = input_idx - (num_prev_esc * 2) - first_input_char_offset;
-    if (num_prev_esc > 0) {
-      size_type prev_esc_pos = *(next_esc_pos_ptr - 1);
-      // find the previous escape to see if this character is within the escape sequence
-      if (input_idx - prev_esc_pos < 3) {
-        if (input_idx == prev_esc_pos) {  // at a position that needs to be replaced
-          ch = (hex_char_to_byte(d_in_chars[input_idx + 1]) << 4) |
-               hex_char_to_byte(d_in_chars[input_idx + 2]);
-          // previous escape sequence is this position, so the original calculation over-adjusted
-          output_idx += 2;
-        } else {
-          // one of the escape hex digits that has no corresponding character in the output
-          return;
-        }
+template <int num_warps_per_threadblock, int char_block_size>
+__global__ void url_decode_char_counter(char const* const in_chars,
+                                        offset_type const* const in_offsets,
+                                        offset_type* const out_counts,
+                                        size_type const num_strings)
+{
+  constexpr int halo_size = 2;
+  __shared__ char temporary_buffer[num_warps_per_threadblock][char_block_size + halo_size];
+  __shared__ typename cub::WarpReduce<int8_t>::TempStorage cub_storage[num_warps_per_threadblock];
+
+  int const global_thread_id = blockIdx.x * blockDim.x + threadIdx.x;
+  int const global_warp_id   = global_thread_id / cudf::detail::warp_size;
+  int const local_warp_id    = threadIdx.x / cudf::detail::warp_size;
+  int const warp_lane        = threadIdx.x % cudf::detail::warp_size;
+  int const nwarps           = gridDim.x * blockDim.x / cudf::detail::warp_size;
+  char* in_chars_shared      = temporary_buffer[local_warp_id];
+
+  // Loop through strings, and assign each string to a warp.
+  for (size_type row_idx = global_warp_id; row_idx < num_strings; row_idx += nwarps) {
+    auto in_chars_string = in_chars + in_offsets[row_idx];
+    auto string_length   = in_offsets[row_idx + 1] - in_offsets[row_idx];
```

Comment:
Here you could use the `column_device_view` to set the these variables.
```
auto d_str = d_input.element<string_view>(row_idx);
auto const in_chars_string = d_str.data();
auto const string_length = d_str.size_bytes();
```
This code is not checking for null rows. Accessing null entries is UB. You can use the `column_device_view` `is_null(row_idx)` to fix that.

Reply:
Checking nulls are also implemented in https://github.com/rapidsai/cudf/pull/8622/commits/93497ab1aaf3688fdc2cb5fae1eb35de0636dcfd.